File: AlgoExpert/1. Easy/Python/2. Find Closest Value In BST/Find Closest Value In BST[O(log(n))].py
# Average: O(log(n)) time | O(log(n)) space
# Worst: O(n) time | O(n) space

def findClosestValueInBst(tree, target):
	return findClosestValueInBstHelper(tree, target, float("inf"))

def findClosestValueInBstHelper(tree, target, closest):
	if tree is None:
		return closest
	if abs(target - closest) > abs(target - tree.value):
		closest = tree.value
	if target < tree.value:
		return findClosestValueInBstHelper(tree.left, target, closest)
	elif target > tree.value:
		return findClosestValueInBstHelper(tree.right, target, closest)
	else:
	return closest

File: AlgoExpert/1. Easy/Java/2. Find Closest Value In BST/Find Closest Value In BST[O(log(n))].java
class Program {

// Average: O(log(n)) time | O(log(n)) space
// Worst: O(n) time | O(n) space

	public static int findClosestValueInBst(BST tree, int target) {
		return findClosestValueInBst(tree, target, Double.MAX_VALUE);
	}

	public static int findClosestValueInBst(BST tree, int target, double closest){
		if (Math.abs(target - closest) > Math.abs(target - tree.value)) {
			closest = tree.value;
		}
		if (target < tree.value && tree.left != null) {
			return findClosestValueInBst(tree.left, target, closest);
		} else if (target > tree.value && tree.right != null) {
			return findClosestValueInBst(tree.right, target, closest);
		} else {
			return (int) closest;
		}
	}

	static class BST {
		public int value;
		public BST left;
		public BST right;
		public BST(int value) {
			this.value = value;
		}
	}
}